src/iface/mod.rs

```diff
@@ -10,3 +10,21 @@ pub use self::neighbor::Neighbor as Neighbor;
 pub(crate) use self::neighbor::Answer as NeighborAnswer;
 pub use self::neighbor::Cache as NeighborCache;
 pub use self::ethernet::Interface as EthernetInterface;
+pub use self::ethernet::Packet as EthernetPacket;
+
+use Result;
+use wire::{IpRepr, UdpRepr, TcpRepr, Icmpv4Repr};
+use phy::{DeviceCapabilities, ChecksumCapabilities};
+
+pub trait PacketFilter {
+    fn process_udp(&mut self, ip_repr: &IpRepr, repr: &UdpRepr) -> Result<()>;
```

Comment:
We need to decide on one of the three possibilities here:
  * Methods of `PacketFilter` are `#[cfg]`d on the existing `socket-*` features. I am reluctant to go with this because you might want to e.g. set a raw packet filter without having the space overhead of having raw socket code in your firmware.
  * Methods of `PacketFilter` are `#[cfg]`d on new `proto-raw/tcp/udp/etc` features, and `EthernetInterface` is changed to also conditionally invoke these methods when the features are enabled. I am reluctant to go with this because of the combinatorial explosion we already have to deal with, but this will provide the most space savings all around.
  * Methods of `PacketFilter` are always present, and so are `EthernetInterface`'s internal code paths for dispatching to them. I am reluctant to go with this because this means we can't quite cut down e.g. TCP even when there's absolutely no need for it.

Reply:
Hmm… I don't think that variant 2 is worth the additional complexity. We can still implement it later if someone has a need for it. For me, variant 1 sounds most reasonable. However, my use case for smoltcp has all no_std features enabled, so I'm not in a good position to make this decision.

Reply:
Let's go with variant 1 for now, then. Since a `socket-udp` feature would always depend on `proto-udp` we can backwards compatibly update to 2 if it's ever needed.